runtime-sdk/src/module.rs

```diff
@@ -1,153 +1,128 @@
 //! Runtime modules.
-use std::{collections::BTreeMap, fmt::Debug};
+use std::fmt::Debug;
 
 use impl_trait_for_tuples::impl_for_tuples;
 use serde::{de::DeserializeOwned, Serialize};
 
 use oasis_core_runtime::common::cbor;
 
 use crate::{
-    context::{DispatchContext, TxContext},
+    context::Context,
     error, event, modules, storage,
     storage::Store,
     types::{
-        message::MessageEvent,
+        message::MessageResult,
         transaction::{CallResult, Transaction, UnverifiedTransaction},
     },
 };
 
-/// Metadata of a callable method.
-pub struct CallableMethodInfo {
-    /// Method name.
-    pub name: &'static str,
-
-    /// Method handler function.
-    pub handler: fn(&CallableMethodInfo, &mut TxContext<'_, '_>, cbor::Value) -> CallResult,
-}
-
-/// Metadata of a query method.
-pub struct QueryMethodInfo {
-    /// Method name.
-    pub name: &'static str,
-
-    /// Method handler function.
-    pub handler: fn(
-        &QueryMethodInfo,
-        &mut DispatchContext<'_>,
-        cbor::Value,
-    ) -> Result<cbor::Value, error::RuntimeError>,
-}
-
-/// Registry of methods exposed by the modules.
-pub struct MethodRegistry {
-    callable_methods: BTreeMap<&'static str, CallableMethodInfo>,
-    query_methods: BTreeMap<&'static str, QueryMethodInfo>,
+/// Result of invoking the method handler.
+pub enum DispatchResult<B, R> {
+    Handled(R),
+    Unhandled(B),
```

Comment:
:information_source: `B` is used to move the body back to the caller

Reply:
actually how much utility do we get from moving the body/message result into the handler? could consider converting to references while we have our sledgehammers out

Reply:
Yeah wanted to do that as well, but `cbor::from_value` wants to own the value so passing a reference won't work without a clone which I wanted to avoid.

Reply:
o